recon_surf/N4_bias_correct.py

```diff
@@ -177,257 +184,388 @@ def N4correctITK(
-    print("- number fitting levels: {}".format(options.levels))
-    print("- number iterations: {}".format(options.numiter))
-    print("- convergence threshold: {}".format(options.thres))
-    print("- skipwm: {}".format(bool(options.skipwm)))
-    if options.tal:
-        print("- talairach: {}".format(options.tal))
-    print("- threads: {}".format(options.threads))
+    LOGLEVEL = (logging.WARNING, logging.INFO, logging.DEBUG)
+    FORMAT = "" if options.verbosity < 0 else "%(levelname)s (%(module)s:%(lineno)s): "
+    FORMAT += "%(message)s"
+    logging.basicConfig(stream=sys.stdout, format=FORMAT)
+    logging.getLogger().setLevel(LOGLEVEL[abs(options.verbosity)])
+    logger = logging.getLogger(__name__)
+    print_options(vars(options))
 
     # set number of threads
     sitk.ProcessObject.SetGlobalDefaultNumberOfThreads(options.threads)
 
     # read image (only nii supported) and convert to float32
-    print("\nreading {}".format(options.invol))
-    # itkimage = sitk.ReadImage(options.invol, sitk.sitkFloat32)
-    itkimage, header = iio.readITKimage(
+    logger.debug(f"reading input volume {options.invol}")
+    # itk_image = sitk.ReadImage(options.invol, sitk.sitkFloat32)
+    itk_image, image_header = iio.readITKimage(
         options.invol, sitk.sitkFloat32, with_header=True
     )
 
-    # if talaraich is passed
-    talorig = None
-    if options.tal:
-        taltrans = readTalairachXFM(options.tal)
-        talorig = getTalOriginVox(taltrans, itkimage)
-
     # read mask (as uchar)
-    if options.mask:
-        itkmask = iio.readITKimage(options.mask, sitk.sitkUInt8)
+    has_mask = bool(options.mask)
+    if has_mask:
+        logger.debug(f"reading mask {options.mask}")
+        itk_mask: sitk.Image = iio.readITKimage(
+            options.mask,
+            sitk.sitkUInt8,
+            with_header=False
+        )
+        # binarize mask
+        itk_mask = cast(sitk.Image, itk_mask > 0)
     else:
-        itkmask = None
+        logger.debug("generate mask with otsu (0, 1, 200)")
+        itk_mask = sitk.OtsuThreshold(itk_image, 0, 1, 200)
+        logger.info("- mask: default (otsu)")
+        itk_mask.CopyInformation(itk_image)
 
     # call N4 correct
-    print("\nexecuting N4 correction ...")
-    itkcorrected = N4correctITK(
-        itkimage,
-        itkmask,
+    logger.info("executing N4 correction ...")
+    itk_bfcorr_image = itk_n4_bfcorrection(
+        itk_image,
+        itk_mask,
         options.shrink,
         options.levels,
         options.numiter,
         options.thres,
-        rescale=options.skipwm,
     )
 
-    if not options.skipwm:
-        print("\nnormalize WM to 110")
-        itkcorrected = normalizeWM(itkcorrected, itkmask, centroid=talorig)
+    target_wm = 110.
+
+    if options.skipwm:
+        logger.info("Skipping WM normalization, ignoring talairach and aseg inputs")
+        # normalize to average input intensity
+        m_bf_img = get_image_mean(itk_bfcorr_image)
```

Comment:
Maybe here we should get the means in the mask (if it is passed) as we do not care much about the intensities outside the brain.

Reply:
I think it probably makes little difference, but we can certainly do that...